### PR TITLE
[solax] Cloud connection support

### PR DESCRIPTION
**Pull request by theater**
Besides the direct connection to the local wi-fi module that we started to support with OpenHAB 4.1, now the binding will support also a new type - cloud connect. The documentation for the API can be obtained from the Solax web portal even though it's not very thorough.

========

Implemented a new thing type - cloud connect inverter.

Refactored the packages, model and classes to reflect that now we have two different ways to retrieve the data.

The common logic is extracted in the AbstractSolaxHandler.
The handler now requires two method implementations with the specific logics for the Local and Cloud connection:
* createConnector() - specific for the two different types of HTTP connection. 
*  updateFromData() - method that maps from the response to the specific thing type channels.

README is updated with the new type and the examples are up to date. Made a small corrections to other parts in the README.

I'm not sure if the thing type is new if I need to create an update instructions. I would expect to NOT need it.

### CODE REVIEW

**Comment by theater** [comment]
Depends on #16248 and has to be rebased on it once it's merged into main

**Comment by theater** [comment]
I have rebased everything on top of the change from the other PR which resulted in a very unpleasant fix of conflicts.
Now the PR is ready for a review. 

Hope someone finds time to review it before any other changes in main branch happen, so I don't have to do this again. It's very time consuming due to a lot of refactoring (including a package refactoring) in this PR.

Best regards,
Konstantin

File: bundles/org.openhab.binding.solax/src/main/resources/OH-INF/thing/cloudConnectInverter.xml
Comment (lsiepel) [inline bundles/org.openhab.binding.solax/src/main/resources/OH-INF/thing/cloudConnectInverter.xml:111]:
```suggestion
				<context>password</context>
```

File: bundles/org.openhab.binding.solax/src/main/resources/OH-INF/thing/cloudConnectInverter.xml
Comment (lsiepel) [inline bundles/org.openhab.binding.solax/src/main/resources/OH-INF/thing/cloudConnectInverter.xml:95]: Don't think this is needed as this is the first version if this type structure and no thing upgrade instructions are written for that first version.

File: bundles/org.openhab.binding.solax/src/main/resources/OH-INF/thing/channel_types.xml
Comment (lsiepel) [inline bundles/org.openhab.binding.solax/src/main/resources/OH-INF/thing/channel_types.xml:57]: This channel now combines both local and cloud values into one channel-type / list . 

I think there should be two variants of this channel-type, one for cloud, one for local, each holder their own set of the values.

File: bundles/org.openhab.binding.solax/src/test/java/org/openhab/binding/solax/internal/cloud/TestCloudParser.java
Comment (lsiepel) [inline bundles/org.openhab.binding.solax/src/test/java/org/openhab/binding/solax/internal/cloud/TestCloudParser.java:87]: Can these tests be fixed?

**Review by lsiepel — CHANGES_REQUESTED** [review]
Did a full review. Most are minor comments. I also think the properties file lacks some values, so it is good to clean/re-run i18n plugin to be sure all are added correctly.

As this is on top of the previous PR and some conflicts have been resolved, i think it would be wise to spend some extra time on testing as we want to prevent regression.

File: bundles/org.openhab.binding.solax/README.md
Comment (theater) [reply bundles/org.openhab.binding.solax/README.md:119]: It's to be on the safe side so we don't reach it due to OH restarts or refresh commands (not very likely though)... I guess 9 or maybe 10 (because it's more natural value) can be used....

File: bundles/org.openhab.binding.solax/README.md
Comment (theater) [reply bundles/org.openhab.binding.solax/README.md:75]: > Is it worth the effort to introduce this beaking change to only rename the channel ?

It's not a breaking change but rather a fix in the documentation. Initially it was state of charge but has been changed. Wasn't change in the README though...

File: bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/connectivity/CloudHttpConnector.java
Comment (theater) [reply bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/connectivity/CloudHttpConnector.java:59]: Why we need this change? I prefer to keep the logic-related traces in DEBUG severity and the full request/response logs in TRACE severity.

File: bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/connectivity/rawdata/cloud/CloudRawDataBean.java
Comment (theater) [reply bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/connectivity/rawdata/cloud/CloudRawDataBean.java:241]: It's a little bit stupid... they put in the exception field the value "success" if all is good and "error" if there is an error. Sorry if it's misleading but I would say that the API response is a bit stupid... Thing is that sometimes the response may be empty and I want to be sure to check that the exception is != null (therefore !error)

File: bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/handlers/AbstractSolaxHandler.java
Comment (theater) [reply bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/handlers/AbstractSolaxHandler.java:113]: What is the idea here? To bring it offline forcefully with a proper status, so we don't end up in the API reject?

File: bundles/org.openhab.binding.solax/src/main/resources/OH-INF/thing/channel_types.xml
Comment (theater) [reply bundles/org.openhab.binding.solax/src/main/resources/OH-INF/thing/channel_types.xml:57]: It's not really an issue because essentially they're the same but I will make it a separate thing type.

File: bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/handlers/AbstractSolaxHandler.java
Comment (theater) [reply bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/handlers/AbstractSolaxHandler.java:50]: hm... not really. I initially implemented it like this and it's a part of the extract from the old handler but can be immediate. I'll change it.

Comment (theater) [reply bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/handlers/AbstractSolaxHandler.java:50]: Btw I haven't seen any overloaded method without an initial delay. Does that mean I need to set the delay to 0?

File: bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/connectivity/CloudHttpConnector.java
Comment (theater) [reply bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/connectivity/CloudHttpConnector.java:59]: Just saw that actually in the Abstract handler it's a debug... will make this one a debug too....

File: bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/connectivity/CloudHttpConnector.java
Comment (theater) [reply bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/connectivity/CloudHttpConnector.java:39]: Hope I understood you correctly. I have removed the constants and put the string values directly in the "replace" methods in the constructor...

File: bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/connectivity/rawdata/cloud/CloudRawDataBean.java
Comment (theater) [reply bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/connectivity/rawdata/cloud/CloudRawDataBean.java:241]: OK I refactored it a bit and also added a to the unit test some negative scenarios. I hope now it's more clear and robust. The issue is that I remember that there was a situation when the schedule was ending with an exception... hope I catched it now with the various unit tests...

**Comment by theater** [comment]
Hi @lsiepel, thanks very much for the fast review! 

I have fixed most of the comments, for some I left my own comments for clarification.

If you decide to meanwhile merge it, please hold until I do some integration testing of the changes, related to the exception management, so I don't pass any missed runtime exception... maybe I should just catch any exception in the AbstractSolaxHandler.retrieveData() method so I ensure another schedule attempt afterwards...

Cheers,
Konstantin

File: bundles/org.openhab.binding.solax/README.md
Comment (lsiepel) [reply bundles/org.openhab.binding.solax/README.md:120]: In that case i would suggest to set the min to 9 as that is the absolute minimum. And set the default to 30 or so.

File: bundles/org.openhab.binding.solax/README.md
Comment (lsiepel) [reply bundles/org.openhab.binding.solax/README.md:76]: Thanks!

File: bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/handlers/AbstractSolaxHandler.java
Comment (lsiepel) [reply bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/handlers/AbstractSolaxHandler.java]: Yes, set it to 0 = immediate.

File: bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/handlers/AbstractSolaxHandler.java
Comment (lsiepel) [reply bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/handlers/AbstractSolaxHandler.java:113]: Maybe if it is very close to the limit, bring ithe Thing offline until it is below some treshold (prevent flapping). That would mean you have to keep tracking of call timestamp. With 10k entry's that is maybe a but to expensive (performance wise).
It might get complex very fast. It's something to think about. The current solution is good enough to merge.

**Comment by theater** [comment]
Hi. I changed the requested minimum refresh interval to 9s, even though from practical perspective it doesn't make sense to refresh lower than 30-60 seconds, because the cloud API data gets updated from the local module once on each 5 minutes. :)
I checked the connectivity and for a whole night no issues (before when there was an issue it stopped working after an hour or so max), so I would expect also that the exception handling is OK and we're OK to merge now, unless you see anything else to add.

Thanks a lot for the review!

**Review by lsiepel — APPROVED** [review]
Thanks, LGTM

**Pull request merged**
